zebra-state/src/service/chain_tip/tests/prop.rs

```diff
@@ -36,81 +42,215 @@ proptest! {
-                BlockUpdate::NonFinalized(block) => {
-                    let chain_tip = block.clone().map(|block| ChainTipBlock::from(block.0));
-                    chain_tip_sender.set_best_non_finalized_tip(chain_tip.clone());
-                    if let Some(block) = block {
-                        latest_non_finalized_tip = Some((chain_tip, block));
-                        seen_non_finalized_tip = true;
+            }
+
+            // check the results
+            let expected_tip = if seen_non_finalized_tip {
+                latest_non_finalized_tip.clone()
+            } else {
+                latest_finalized_tip.clone()
+            };
+
+            let chain_tip_height = expected_tip
+                .as_ref()
+                .map(|(chain_tip, _block)| chain_tip.height);
+            let expected_height = expected_tip.as_ref().and_then(|(_chain_tip, block)| block.coinbase_height());
+            prop_assert_eq!(latest_chain_tip.best_tip_height(), chain_tip_height);
+            prop_assert_eq!(latest_chain_tip.best_tip_height(), expected_height);
+
+            let chain_tip_hash = expected_tip
+                .as_ref()
+                .map(|(chain_tip, _block)| chain_tip.hash);
+            let expected_hash = expected_tip.as_ref().map(|(_chain_tip, block)| block.hash());
+            prop_assert_eq!(latest_chain_tip.best_tip_hash(), chain_tip_hash);
+            prop_assert_eq!(latest_chain_tip.best_tip_hash(), expected_hash);
+
+            let chain_tip_transaction_ids = expected_tip
+                .as_ref()
+                .map(|(chain_tip, _block)| chain_tip.transaction_hashes.clone())
+                .unwrap_or_else(|| Arc::new([]));
+            let expected_transaction_ids = expected_tip
+                .as_ref()
+                .iter()
+                .flat_map(|(_chain_tip, block)| block.transactions.clone())
+                .map(|transaction| transaction.hash())
+                .collect();
+            prop_assert_eq!(
+                latest_chain_tip.best_tip_mined_transaction_ids(),
+                chain_tip_transaction_ids
+            );
+            prop_assert_eq!(
+                latest_chain_tip.best_tip_mined_transaction_ids(),
+                expected_transaction_ids
+            );
+
+            let old_last_change_hash = chain_tip_change.last_change_hash;
+
+            let new_action =
+                expected_tip.and_then(|(chain_tip, block)| {
+                    if Some(chain_tip.hash) == old_last_change_hash {
+                        // some updates don't do anything, so there's no new action
+                        None
+                    } else if Some(chain_tip.previous_block_hash) != old_last_change_hash ||
+                        NetworkUpgrade::is_activation_height(network, chain_tip.height) {
+                            Some(TipAction::reset_with(block.0.into()))
+                    } else {
+                            Some(TipAction::grow_with(block.0.into()))
```

Comment:
Optional nit: These lines look like they could use another round of `rustfmt`.

Reply:
Done in b0a7d207e.

I'm looking forward to rustfmt getting better macro formatting, or rust-analyzer getting the ability to format regions (rather than whole files).